src/Everscale/Cell.cpp

```diff
@@ -0,0 +1,384 @@
+
+        // Write cells
+        size_t i = cellCount - 1;
+        while (true) {
+            const auto& cell = *reversedCells[i];
+
+            // Write cell data
+            const auto [d1, d2] = cell.getDescriptorBytes();
+            os.push_back(d1);
+            os.push_back(d2);
+            os.insert(os.end(), cell.data.begin(), cell.data.end());
+
+            // Write cell references
+            for (const auto& child : cell.references) {
+                if (child == nullptr) {
+                    break;
+                }
+
+                // Map cell hash to index (which must be presented)
+                const auto it = indices.find(child->hash);
+                assert(it != indices.end());
+
+                encode16BE(cellCount - it->second - 1, os);
+            }
+
+            if (i == 0) {
+                break;
+            } else {
+                --i;
+            }
+        }
+    }
+
+private:
+    // uint16_t will be enough for wallet transactions (e.g. 64k is the size of the whole elector)
+    constexpr static uint8_t REF_SIZE = 2;
+    constexpr static uint8_t OFFSET_SIZE = 2;
+
+    // NOTE: Initialized with header size
+    size_t serializedSize =
+        /*magic*/ sizeof(BOC_MAGIC) +
+        /*ref_size*/ 1 +
+        /*offset_size*/ 1 +
+        /*cell_count*/ REF_SIZE +
+        /*root_count*/ REF_SIZE +
+        /*absent_count*/ REF_SIZE +
+        /*data_size*/ OFFSET_SIZE +
+        /*root_cell_index*/ REF_SIZE;
+
+    uint16_t index = 0;
+    std::map<Cell::CellHash, uint16_t> indices{};
+    std::vector<const Cell* _Nonnull> reversedCells{};
+
+    static void fillContext(const Cell& cell, SerializationContext& ctx) {
+        if (ctx.indices.find(cell.hash) != ctx.indices.end()) {
+            return;
+        }
+
+        for (const auto& ref : cell.references) {
+            if (ref == nullptr) {
+                break;
+            }
+            fillContext(*ref, ctx);
+        }
+
+        ctx.indices.insert(std::make_pair(cell.hash, ctx.index++));
+        ctx.reversedCells.emplace_back(&cell);
+        ctx.serializedSize += cell.serializedSize();
+    }
+};
+
+void Cell::serialize(Data& os) const {
+    assert(finalized);
+    const auto ctx = SerializationContext::build(*this);
+    ctx.encode(os);
+}
+
+void Cell::finalize() {
+    if (finalized) {
+        return;
+    }
+
+    if (bitLen > Cell::MAX_BITS || refCount > Cell::MAX_REFS) {
+        throw std::invalid_argument("invalid cell");
+    }
+
+    // Finalize child cells and update current cell depth
+    // NOTE: Use before context creation to reduce stack size
+    for (const auto& ref : references) {
+        if (ref == nullptr) {
+            break;
+        }
+        ref->finalize();
+        depth = std::max(depth, static_cast<uint16_t>(ref->depth + 1));
+    }
+
+    // Compute cell hash
+    // NOTE: raw context is used here to reduce allocations
+    SHA256_CTX context{};
```

Comment:
Please use Hash::sha256 if possible instead of handling those lower-level things manually

Reply:
Should we use this even if it involves unnecessary allocations in that case? `Hash::sha256` requires continuous memory, but the normalized representation is computed here from multiple parts